### PR TITLE
Add: Train ViT on small datasets

### PR DESCRIPTION
**Pull request by ariG23498**
In this example I implement [Vision Transformer for Small-Size Datasets](https://arxiv.org/abs/2112.13492v1) by Lee et. al. In the paper the authors propose Shifted Patch Tokenization (SPT) and Locality Self Attention (LSA) to improve the locality inductive bias of ViTs. With LSA and SPT my mini ViT gained **~4%** top-1 test accuracy.

In the example I have subclassed the `MultiHeadAttention` layer of `tf.keras` to implemet LSA. SPT is also wrapped as a `keras.layers.Layer` to plug and play. I hope the community will find this example helpful.

[Colab Notebook](https://colab.research.google.com/drive/1HEIsLhJkJ5PdFKvZ-CEvsYfN0vLgE07h?usp=sharing)

### CODE REVIEW

**Review by aanna0701 — APPROVED** [review]
First of all, thank you very much for your interest in my research.

Both SPT and LSA have been perfectly implemented.

File: examples/vision/vit_small_ds.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+            attention_scores, training=training
+        )
+        attention_output = tf.einsum(
+            self._combine_equation, attention_scores_dropout, value
+        )
+        return attention_output, attention_scores
+
+
+"""
+## Implement the mlp
```
Comment (fchollet) [inline examples/vision/vit_small_ds.py:376]: MLP

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: examples/vision/vit_small_ds.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
+"""
+Title: Train a Vision Transformer on Smaller Datasets
+Author: [Aritra Roy Gosthipaty](https://twitter.com/ariG23498)
+Date created: 2022/01/07
+Last modified: 2022/01/07
+Description: Training a ViT from scratch on smaller datasets with shifted patch tokenization and locality self attention.
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply examples/vision/vit_small_ds.py:6]: This example provides two novel ideas:
- Shifted Patch Tokenization: A tokenization scheme that helps capture the spatial relationship. Here we do not use a convolution as is used in MobileViT. The trick here is to use a concatenation of shifted images and tokenizing them.
- Locality Self Attention: The transformer architecture does not change as opposed to MobileViT. The authors introduce a diagonal mask and a learnable temperature term.

File: examples/vision/vit_small_ds.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+A snippet from the paper:
+
+"
+According to DeiT, various techniques are required to effectively train ViTs. Thus, we
+applied data augmentations such as CutMix, Mixup, Auto Augment, Repeated Augment to all
+models.
+"
+
+In this example, we will focus solely on the novelty of the approach and not on
+reproducing the paper results. This is the reason why I have not used the mentioned data
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply examples/vision/vit_small_ds.py:119]: Noted!

File: examples/vision/vit_small_ds.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+for row in range(n):
+    for col in range(n):
+        plt.subplot(n, n, count)
+        count = count + 1
+        image = tf.reshape(patch[row][col], (PATCH_SIZE, PATCH_SIZE, 3))
+        plt.imshow(image)
+        plt.axis("off")
+plt.show()
+
+# SPT: This layer takes the image, shifts it diagonally and then
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply examples/vision/vit_small_ds.py:275]: Noted!

**Comment by ariG23498** [comment]
@fchollet thank you for the review. I have provided the changes.

**Review by fchollet — APPROVED** [review]
Thanks for the update! I pushed some copyedits, please pull them first.

At this stage, you can add the autogenerated files.

**Comment by ariG23498** [comment]
@fchollet I have added the autogenerated files.

Note: The description was more that 100 characters, which resulted the generation to error out. I have bypassed the generation code and kept the description as is. Let me know if that is a problem.
